README.md
# FPNN 介绍

**FPNN一句话总结：**

+ **客户端**：一个API解决所有操作
+ **服务器**：一个类继承解决所有 RPC、同步/异步、编码/解码、TCP/UDP 问题

## 一、开发背景

没有人想重复制造轮子。

但，当所有的轮都不好使的时候，怎么办？

鉴于之前的项目使用Thrift开发遇到了太多的坑，耗费大量的资源解决Thrift本身问题。  
基于对公司基础架构工具的准备及技术积累，结合之前对ICE、ACE、Codra等知名RPC框架的使用经验，于是决定开发一个更适合公司业务及基础架构的RPC框架。

轮子没有最好，只有最合适。

RPC框架也一样。

## 二、FPNN框架提供功能

* 支持 IPv4
* 支持 IPv6
* 支持 TCP 二进制私有协议
* 支持 UDP 二进制私有协议
* 支持可靠 UDP 链接
* 支持可靠/不可靠 UDP 数据混合发送
* 支持 HTTP 1.1 (选择性支持)
* 支持 WebSocket
* 支持 SSL/TLS
* 支持 msgpack 编码
* 支持 Json 格式
* 支持 可选参数
* 支持 不定类型参数
* 支持 不定长度不定类型参数
* 支持 接口灰度兼容
* 支持 TCP 二进制私有协议和 HTTP Json 格式互转

* 支持 Server Push
* 支持 异步操作
* 支持 同步操作
* 支持 Lambda 函数
* 支持 动态调整系统级参数
* 支持 动态调整框架级参数
* 支持 实时查看服务运行状态
* 支持 实时查看各参数状态
* 支持 failover
* 支持 统一 Log 汇总
* 支持 优雅退出
* 支持 统一处理异常
* 支持 同一端口多种协议(TCP/HTTP/WebSocket)
* 支持 应答提前返回
* 支持 应答异步返回
* 支持 请求响应时间统计
* 支持 QPS 统计
* 支持 慢请求统计
* 支持 优先执行系统内置命令

* 支持 AES 加密
* 支持 ECDH 秘钥交换
* 支持 128 位或 256 位秘钥
* 支持 IP 白名单
* 支持 IP 段白名单
* 支持 访问用户自定义接口时的加密限制
* 支持 访问用户自定义接口时的内网限制

在 FPNN 技术生态中，提供以下额外功能

* 支持 集群注册
* 支持 集群管理
* 支持 数据协调
* 支持 数据同步
* 支持 大规模分布式测试部署、监控、协调
* 支持 FPNN 集群透明代理
* 支持 MySQL 透明代理
* 支持 数据行级缓存

## 三、FPNN框架性能概览

压力测试(v1.0.0)：

**TCP 数据概要**

* 同一局域网

	| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） |
	|-----|---------|-----------|---------|-----|------------------|
	| AWS m5.xlarge | 4 | 16 | 1000 | 49,708 | 335 |
	|---|---|---|---|---|---|
	| AWS m5.xlarge | 4 | 16 | 130 | 227,919 | 12,854 |
	| AWS m5.xlarge | 4 | 16 | 1,500 | 148,959 | 10,403 |
	| AWS m5.xlarge | 4 | 16 | 2,000 | 99,552 | 356 |
	| AWS m5.xlarge | 4 | 16 | 3,000 | 149,615 | 27,456 |

* 洲际传输：德国法兰克福 到 美国西部俄勒冈

	| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） | ping/2 (msec) |
	|-----|---------|-----------|---------|-----|-------------------|---------------|
	| AWS m5.xlarge | 4 | 16 | 10 | 17,921 | 147,825 | 137 |
	|---|---|---|---|---|----|---|
	| AWS m5.xlarge | 4 | 16 | 190 | 338,601 | 151,772 | 139 |
	| AWS m5.xlarge | 4 | 16 | 700 | 339,240 | 183,541 | 136 ~ 137 |
	| AWS m5.xlarge | 4 | 16 | 3,200 | 312,073 | 256,980 | 139 ~ 141 |
	| AWS m5.xlarge | 4 | 16 | 6,000 | 299,175 | 346,927 | 136 ~ 137 |

**UDP 数据概要**

* 同一局域网

	| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） |
	|-----|---------|-----------|---------|-----|------------------|
	| AWS m5.xlarge | 4 | 16 | 10 | 17,628 | 199 |
	|---|---|---|---|---|---|
	| AWS m5.xlarge | 4 | 16 | 70 | 123,009 | 1,174 |
	| AWS m5.xlarge | 4 | 16 | 80 | 122,641 | 24,623 |
	| AWS m5.xlarge | 4 | 16 | 800 | 73,030 | 1,306 |
	| AWS m5.xlarge | 4 | 16 | 200 | 95,894 | 400 |
	| AWS m5.xlarge | 4 | 16 | 1,000 | 49,590 | 3,832 |

* 洲际传输：德国法兰克福 到 美国西部俄勒冈

	| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） | ping/2 (msec) |
	|-----|---------|-----------|---------|-----|-------------------|---------------|
	| AWS m5.xlarge | 4 | 16 | 100 | 48,493 | 138,859 | 138 |
	|---|---|---|---|---|----|---|
	| AWS m5.xlarge | 4 | 16 | 60 | 107,528 | 139,968 | 140 |
	| AWS m5.xlarge | 4 | 16 | 800 | 79,503 | 139,974 | 138 ~ 139 |


海量链接(v1.0.0)：

**TCP 数据概要**

| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） |
|-----|---------|-----------|---------|-----|------------------|
| AWS m5.2xlarge | 8 | 32 | 2,040,000 | 81,351 | 446 |
| AWS m5.2xlarge | 8 | 32 | 2,040,000 | 137,294 | 4,985 |
| AWS m5.2xlarge | 8 | 32 | 2,040,000 | 179,794 | 11,345 |

**UDP 数据概要**

| 机型 | 虚拟 CPU | 内存（GB） | 链接数量 | QPS | 平均响应时间（usec） |
|-----|---------|-----------|---------|-----|------------------|
| AWS m5.2xlarge | 8 | 32 | 12,200 | 5,957 | 11,480 |
| AWS m5.2xlarge | 8 | 32 | 19,000 | 721 | 1,857 |

具体性能介绍及更多数据请参见 [FPNN 性能报告](doc/zh-cn/fpnn-performance.md)


## 四、使用

1. 环境需求

	| 操作系统 | 编译器 |
	|---------|-------|
	| CentOS 7 | GCC/G++ 4.8.5 |
	| CentOS 8 | GCC/G++ 8 |
	| Ubuntu 20 | GCC/G++ 9 |
	| MacOS 11.5 | XCode/Apple clang 12 |

1. 第三方库依赖

	+ gcc
	+ g++
	+ libcurl
	+ tcmalloc
	+ openssl

1. 编译安装FPNN框架

	在项目根目录下执行 `make` 即可。

	详细可参见 [FPNN 安装与集成](doc/zh-cn/fpnn-install.md)

1. 使用 FPNN 框架开发

	请参见

	1. [FPNN 服务端基础使用说明](doc/zh-cn/fpnn-server-basic-tutorial.md)
	1. [FPNN 服务端高级使用说明](doc/zh-cn/fpnn-server-advanced-tutorial.md)
	1. [FPNN 客户端基础使用说明](doc/zh-cn/fpnn-client-basic-tutorial.md)
	1. [FPNN 客户端高级使用说明](doc/zh-cn/fpnn-client-advanced-tutorial.md)

1. 注意事项 & 问题排查

	注意事项请参见 [FPNN 注意事项](doc/zh-cn/fpnn-notices.md)

	问题排查请参见 [FPNN 问题排查](doc/zh-cn/fpnn-troubleshooting.md)

1. 最佳实践

	1. [FPNN 最佳实践](doc/zh-cn/fpnn-best-practice.md)
	1. [FPNN 服务治理与服务网格](doc/zh-cn/fpnn-SOA-governance-service-mesh.md)

## 五、全部文档索引

[中文文档索引](doc/zh-cn/README.md)
 


changes.md
### v.1.3.1

#### 修复

* 满尺寸的UDP包，因为重发默认是组装包格式，所以因剩余空间小2个字节，导致无法重发的问题。


-----------

### v.1.3.0

#### 增加

* 原始 TCP 客户端支持
	+ 接收和处理顺序保证
	+ 无协议处理和序列化，须自行定义
	+ 用于对接其他 RPC 协议
	+ 用于对接原始数据流
* 原始 UDP 客户端支持
	+ 无 ARQ 支持
		+ 无自动分片和自动重组支持
		+ 无顺序重整支持
		+ 无去重支持
		+ 无重传支持
	+ 无协议处理和序列化，须自行定义
	+ 用于对接其他机遇 UDP 的协议

#### 修复

* Base/FileSystemUtil 模块，FileAttrs 结构中，文件尺寸和时间相关字段，改为 64 位长度。


-----------


### v.1.2.2

#### 修复

* WebSocket 分帧支持处理笔误导致分帧支持失效的问题。


-----------


### v.1.2.1

#### 修复

* UDP 四元组重入问题
* MacOS M1 芯片和ARM服务器兼容问题


-----------


### v.1.2.0

#### 增加

* 增加 CUDA 支持和 GPU 监控
* extends/FPZKClient 增加 GPU 信息同步功能
* extends/FPZKClient 增加 自定义数据 同步功能
* 新增 UDP v2 模块
	+ 兼容 UDP v1 协议和处理
	+ 支持加密（整体进行包加密） & 强化加密（数据部分进行流加密）
	+ 支持提前拆解包：普通包、复合包、组装包
	+ 支持重发包合并与包组合（组装包）
	+ 废弃失效/超时的重发包
		* 例外：首包永不被废弃
		* 例外：启用强化加密的数据包不会被废弃
	+ 废弃超时的分片数据：分片数据取消通知

* infos 接口，线程池部分，增加 ArraySize 显示线程池阵列数量

#### 变更

* MachineStatus::getConnectionCount() 变更为 MachineStatus::getInusedSocketCount()
* extends/FPZKClient 将 connCount 及 connNum 拆分为 tcpCount 及 tcpNum 与 udpCount 及 udpNum
* extends/FPZKClient 服务变动通知相关 API 改动，将区分变动的集群和无效的（没有可用服务实例的）集群
* TCP & UDP 服务器模块、ClientEngine 模块的 I/O、worker、duplex 线程池配置修改：
	+ 如果初始线程数量小于 CPU 核心数，则线程池阵列数量为初始线程数量
	+ 如果初始线程数量大于 CPU 核心数，则线程池阵列数量为 CPU 核心数量


#### 修复

* FPTimer 任务过多，唤醒操作会卡死问题


#### 移除

* UDP v1 实现相关代码

base/md5.c
/*
 * RFC 1321 compliant MD5 implementation
 *
 * Copyright (C) 2001-2003	 Christophe Devine
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "md5.h"
#include <string.h>

#define GET_UINT32(n,b,i)				\
{							\
	(n) = ( (uint32_t) (b)[(i)] )			\
		| ( (uint32_t) (b)[(i) + 1] <<	8 )	\
		| ( (uint32_t) (b)[(i) + 2] << 16 )	\
		| ( (uint32_t) (b)[(i) + 3] << 24 );	\
}

#define PUT_UINT32(n,b,i)				\
{							\
	(b)[(i)	   ] = (unsigned char) ( (n)	   );	\
	(b)[(i) + 1] = (unsigned char) ( (n) >>	 8 );	\
	(b)[(i) + 2] = (unsigned char) ( (n) >> 16 );	\
	(b)[(i) + 3] = (unsigned char) ( (n) >> 24 );	\
}

void md5_start(md5_context * ctx)
{
	ctx->total[0] = 0;
	ctx->total[1] = 0;

	ctx->state[0] = 0x67452301;
	ctx->state[1] = 0xEFCDAB89;
	ctx->state[2] = 0x98BADCFE;
	ctx->state[3] = 0x10325476;
}

static void md5_process(md5_context * ctx, unsigned char data[64])
{
	uint32_t X[16], A, B, C, D;

	GET_UINT32(X[0], data, 0);
	GET_UINT32(X[1], data, 4);
	GET_UINT32(X[2], data, 8);
	GET_UINT32(X[3], data, 12);
	GET_UINT32(X[4], data, 16);
	GET_UINT32(X[5], data, 20);
	GET_UINT32(X[6], data, 24);
	GET_UINT32(X[7], data, 28);
	GET_UINT32(X[8], data, 32);
	GET_UINT32(X[9], data, 36);
	GET_UINT32(X[10], data, 40);
	GET_UINT32(X[11], data, 44);
	GET_UINT32(X[12], data, 48);
	GET_UINT32(X[13], data, 52);
	GET_UINT32(X[14], data, 56);
	GET_UINT32(X[15], data, 60);

#define S(x,n) 	((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

#define P(a,b,c,d,k,s,t)  do { a += F(b,c,d) + X[k] + t; a = S(a,s) + b; } while(0)

	A = ctx->state[0];
	B = ctx->state[1];
	C = ctx->state[2];
	D = ctx->state[3];

#define F(x,y,z) (z ^ (x & (y ^ z)))
	P(A, B, C, D, 0, 7, 0xD76AA478);
	P(D, A, B, C, 1, 12, 0xE8C7B756);
	P(C, D, A, B, 2, 17, 0x242070DB);
	P(B, C, D, A, 3, 22, 0xC1BDCEEE);
	P(A, B, C, D, 4, 7, 0xF57C0FAF);
	P(D, A, B, C, 5, 12, 0x4787C62A);
	P(C, D, A, B, 6, 17, 0xA8304613);
	P(B, C, D, A, 7, 22, 0xFD469501);
	P(A, B, C, D, 8, 7, 0x698098D8);
	P(D, A, B, C, 9, 12, 0x8B44F7AF);
	P(C, D, A, B, 10, 17, 0xFFFF5BB1);
	P(B, C, D, A, 11, 22, 0x895CD7BE);
	P(A, B, C, D, 12, 7, 0x6B901122);
	P(D, A, B, C, 13, 12, 0xFD987193);
	P(C, D, A, B, 14, 17, 0xA679438E);
	P(B, C, D, A, 15, 22, 0x49B40821);
#undef F

#define F(x,y,z) (y ^ (z & (x ^ y)))
	P(A, B, C, D, 1, 5, 0xF61E2562);
	P(D, A, B, C, 6, 9, 0xC040B340);
	P(C, D, A, B, 11, 14, 0x265E5A51);
	P(B, C, D, A, 0, 20, 0xE9B6C7AA);
	P(A, B, C, D, 5, 5, 0xD62F105D);
	P(D, A, B, C, 10, 9, 0x02441453);
	P(C, D, A, B, 15, 14, 0xD8A1E681);
	P(B, C, D, A, 4, 20, 0xE7D3FBC8);
	P(A, B, C, D, 9, 5, 0x21E1CDE6);
	P(D, A, B, C, 14, 9, 0xC33707D6);
	P(C, D, A, B, 3, 14, 0xF4D50D87);
	P(B, C, D, A, 8, 20, 0x455A14ED);
	P(A, B, C, D, 13, 5, 0xA9E3E905);
	P(D, A, B, C, 2, 9, 0xFCEFA3F8);
	P(C, D, A, B, 7, 14, 0x676F02D9);
	P(B, C, D, A, 12, 20, 0x8D2A4C8A);
#undef F

#define F(x,y,z) (x ^ y ^ z)
	P(A, B, C, D, 5, 4, 0xFFFA3942);
	P(D, A, B, C, 8, 11, 0x8771F681);
	P(C, D, A, B, 11, 16, 0x6D9D6122);
	P(B, C, D, A, 14, 23, 0xFDE5380C);
	P(A, B, C, D, 1, 4, 0xA4BEEA44);
	P(D, A, B, C, 4, 11, 0x4BDECFA9);
	P(C, D, A, B, 7, 16, 0xF6BB4B60);
	P(B, C, D, A, 10, 23, 0xBEBFBC70);
	P(A, B, C, D, 13, 4, 0x289B7EC6);
	P(D, A, B, C, 0, 11, 0xEAA127FA);
	P(C, D, A, B, 3, 16, 0xD4EF3085);
	P(B, C, D, A, 6, 23, 0x04881D05);
	P(A, B, C, D, 9, 4, 0xD9D4D039);
	P(D, A, B, C, 12, 11, 0xE6DB99E5);
	P(C, D, A, B, 15, 16, 0x1FA27CF8);
	P(B, C, D, A, 2, 23, 0xC4AC5665);
#undef F

#define F(x,y,z) (y ^ (x | ~z))
	P(A, B, C, D, 0, 6, 0xF4292244);
	P(D, A, B, C, 7, 10, 0x432AFF97);
	P(C, D, A, B, 14, 15, 0xAB9423A7);
	P(B, C, D, A, 5, 21, 0xFC93A039);
	P(A, B, C, D, 12, 6, 0x655B59C3);
	P(D, A, B, C, 3, 10, 0x8F0CCC92);
	P(C, D, A, B, 10, 15, 0xFFEFF47D);
	P(B, C, D, A, 1, 21, 0x85845DD1);
	P(A, B, C, D, 8, 6, 0x6FA87E4F);
	P(D, A, B, C, 15, 10, 0xFE2CE6E0);
	P(C, D, A, B, 6, 15, 0xA3014314);
	P(B, C, D, A, 13, 21, 0x4E0811A1);
	P(A, B, C, D, 4, 6, 0xF7537E82);
	P(D, A, B, C, 11, 10, 0xBD3AF235);
	P(C, D, A, B, 2, 15, 0x2AD7D2BB);
	P(B, C, D, A, 9, 21, 0xEB86D391);
#undef F

	ctx->state[0] += A;
	ctx->state[1] += B;
	ctx->state[2] += C;
	ctx->state[3] += D;
}

void md5_update(md5_context * ctx, const void *input, size_t length)
{
	uint32_t left, fill, len;

	if (!length)
		return;

	left = ctx->total[0] & 0x3F;
	fill = 64 - left;

	len = length;
#if UINT32_MAX < SIZE_MAX
	if (len < length)
		ctx->total[1] += (length >> 32);
#endif

	ctx->total[0] += len;
	ctx->total[0] &= 0xFFFFFFFF;
	if (ctx->total[0] < len)
		ctx->total[1]++;

	if (left && length >= fill) {
		memcpy((void *) (ctx->buffer + left), (void *) input, fill);
		md5_process(ctx, ctx->buffer);
		length -= fill;
		input = (char *)input + fill;
		left = 0;
	}

	while (length >= 64) {
		md5_process(ctx, (unsigned char *)input);
		length -= 64;
		input = (char *)input + 64;
	}

	if (length) {
		memcpy((ctx->buffer + left), input, length);
	}
}

static unsigned char md5_padding[64] = {
     0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

inline void md5_get(md5_context *ctx, unsigned char digest[16])
{
	PUT_UINT32(ctx->state[0], digest, 0);
	PUT_UINT32(ctx->state[1], digest, 4);
	PUT_UINT32(ctx->state[2], digest, 8);
	PUT_UINT32(ctx->state[3], digest, 12);
}

void md5_finish(md5_context * ctx, unsigned char digest[16])
{
	uint32_t last, padn;
	uint32_t high, low;
	unsigned char msglen[8];

	high = (ctx->total[0] >> 29) | (ctx->total[1] << 3);
	low = (ctx->total[0] << 3);

	PUT_UINT32(low, msglen, 0);
	PUT_UINT32(high, msglen, 4);

	last = ctx->total[0] & 0x3F;
	padn = (last < 56) ? (56 - last) : (120 - last);

	md5_update(ctx, md5_padding, padn);
	md5_update(ctx, msglen, 8);

	md5_get(ctx, digest);
}

void md5_checksum(unsigned char digest[16], const void *input, size_t length)
{
	md5_context ctx;

	md5_start(&ctx);
	md5_update(&ctx, input, length);
	md5_finish(&ctx, digest);
}


#ifdef TEST_MD5

#include <stdlib.h>
#include <stdio.h>

/*
 * those are the standard RFC 1321 test vectors
 */

static char *msg[] = {
	"",
	"a",
	"abc",
	"message digest",
	"abcdefghijklmnopqrstuvwxyz",
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789",
	"12345678901234567890123456789012345678901234567890123456789012345678901234567890"
};

static char *val[] = {
	"d41d8cd98f00b204e9800998ecf8427e",
	"0cc175b9c0f1b6a831c399e269772661",
	"900150983cd24fb0d6963f7d28e17f72",
	"f96b697d7cb7938d525a2f31aaf161d0",
	"c3fcd3d76192e4007dfb496cca67e13b",
	"d174ab98d277d9f5a5611c2c9f419d9f",
	"57edf4a22be3c955ac49da2e2107b67a"
};

int main(int argc, char *argv[])
{
	FILE *f;
	int i, j;
	char output[33];
	md5_context ctx;
	unsigned char buf[1000];
	unsigned char md5sum[16];

	if (argc < 2)
	{
		printf("\n MD5 Validation Tests:\n\n");

		for (i = 0; i < 7; i++) {
			printf(" Test %d ", i + 1);

			md5_start(&ctx);
			md5_update(&ctx, (unsigned char *) msg[i], strlen(msg[i]));
			md5_finish(&ctx, md5sum);

			for (j = 0; j < 16; j++) {
				sprintf(output + j * 2, "%02x", md5sum[j]);
			}

			if (memcmp(output, val[i], 32)) {
				printf("failed!\n");
				return 1;
			}

			printf("passed.\n");
		}
		printf("\n");
	}
	else
	{
		if (!(f = fopen(argv[1], "rb"))) {
			perror("fopen");
			return 1;
		}

		md5_start(&ctx);

		while ((i = fread(buf, 1, sizeof(buf), f)) > 0) {
			md5_update(&ctx, buf, i);
		}

		md5_finish(&ctx, md5sum);

		for (j = 0; j < 16; j++) {
			printf("%02x", md5sum[j]);
		}

		printf("  %s\n", argv[1]);
	}

	return 0;
}

#endif
